frappe/api/v2.py

```diff
@@ -68,17 +68,42 @@ def read_doc(doctype: str, name: str):
 	doc = frappe.get_doc(doctype, name)
 	doc.check_permission("read")
 	doc.apply_fieldlevel_read_permissions()
-	return doc
+	return doc.as_dict()
 
 
 def document_list(doctype: str):
-	if frappe.form_dict.get("fields"):
-		frappe.form_dict["fields"] = json.loads(frappe.form_dict["fields"])
-
-	# set limit of records for frappe.get_list
-	frappe.form_dict.limit_page_length = frappe.form_dict.limit or 20
-	# evaluate frappe.get_list
-	return frappe.call(frappe.client.get_list, doctype, **frappe.form_dict)
+	from frappe.model.base_document import get_controller
+
+	args = frappe.form_dict
+
+	fields = frappe.parse_json(args.get("fields", None))
+	filters = frappe.parse_json(args.get("filters", None))
+	order_by = args.get("order_by", None)
+	start = cint(args.get("start", 0))
+	limit = cint(args.get("limit", 20))
+	group_by = args.get("group_by", None)
+	debug = args.get("debug", False)
+	as_dict = args.get("as_dict", True)
+
+	query = frappe.qb.get_query(
+		table=doctype,
+		fields=fields,
+		filters=filters,
+		order_by=order_by,
+		offset=start,
+		limit=limit + 1,
+		group_by=group_by,
+		ignore_permissions=False,
+	)
+	controller = get_controller(doctype)
+	if hasattr(controller, "get_list"):
+		return_value = controller.get_list(query)
+		if return_value is not None:
+			query = return_value
```

Comment:
This is a replacement for `permission_query_conditions`, right? But what happens to the ones already defined (they're used for enforcing authorized access everywhere)

can we somehow shove those partial queries in the generated query? :thinking:

Reply:
Granted we won't be able to support all current ones but at least they'll break and not silently skip perms.

Reply:
Added support for `permission_query_conditions`